### PR TITLE
Add interactive command darwin framework tool

### PR DESCRIPTION
**Pull request by krypton36**
#### Problem
Darwin-framework-tool does not have an interactive mode to better validate subscriptions, PASE only with following commands, and server behavior. 

#### Change overview
- Add ability to run interactive mode in Darwin-framework-tool

#### Testing
- Generated code
- Compiled
- Ran Interactive 
- Ran pairing in interactive
- Ran commands in interactive. 
- Used quit() to leave interactive.

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]
Fast tracking platform changes

**Review by woody-apple — APPROVED** [review]
Fast tracking platform changes

File: examples/darwin-framework-tool/commands/clusters/ReportCommandBridge.h [hunk 1]
```diff
@@ -132,35 +132,45 @@ class SubscribeAttribute : public ModelCommand {
...
     uint16_t mMaxInterval;
     bool mWait;
 
+    void Shutdown() override
+    {
+        mSubscriptionEstablished = NO;
+        ModelCommand::Shutdown();
+    }
+
+    virtual bool DeferInteractiveCleanup() override { return mSubscriptionEstablished; }
```
Comment (bzbarsky-apple) [inline examples/darwin-framework-tool/commands/clusters/ReportCommandBridge.h:172]:
```suggestion
    bool DeferInteractiveCleanup() override { return mSubscriptionEstablished; }
```

File: examples/darwin-framework-tool/commands/clusters/ReportCommandBridge.h [hunk 2]
```diff
@@ -132,35 +132,45 @@ class SubscribeAttribute : public ModelCommand {
...
 
 protected:
     chip::Optional<bool> mKeepSubscriptions;
     chip::Optional<bool> mFabricFiltered;
+    bool mSubscriptionEstablished = NO;
     uint16_t mMinInterval;
     uint16_t mMaxInterval;
     bool mWait;
 
+    void Shutdown() override
```
Comment (bzbarsky-apple) [inline examples/darwin-framework-tool/commands/clusters/ReportCommandBridge.h:166]: Followup: our Cleanup should cancel the subscription, I would think, but right now it doesn't do that.

File: examples/darwin-framework-tool/commands/common/CHIPCommandBridge.mm [hunk 3]
```diff
@@ -76,15 +102,20 @@
...
 }
 
+CHIP_ERROR CHIPCommandBridge::MaybeTearDownStack()
+{
+    CHIP_ERROR err;
+    if (IsInteractive()) {
+        return CHIP_NO_ERROR;
+    }
+    err = ShutdownCommissioner();
+    ReturnErrorOnFailure(chip::DeviceLayer::PlatformMgr().StopEventLoopTask());
```
Comment (bzbarsky-apple) [inline examples/darwin-framework-tool/commands/common/CHIPCommandBridge.mm:115]: This doesn't make sense here, fwiw.  It'll happen automatically in commissioner shutdown.

I am pretty sure the `StartEventLoopTask` call in StartWaiting can also go away.

File: examples/darwin-framework-tool/commands/common/CHIPCommandBridge.mm [hunk 4]
```diff
@@ -76,15 +102,20 @@
...
     return CHIP_NO_ERROR;
 }
 
+CHIP_ERROR CHIPCommandBridge::MaybeTearDownStack()
+{
+    CHIP_ERROR err;
+    if (IsInteractive()) {
+        return CHIP_NO_ERROR;
+    }
+    err = ShutdownCommissioner();
```
Comment (bzbarsky-apple) [inline examples/darwin-framework-tool/commands/common/CHIPCommandBridge.mm:114]: What does the shutdown in interactive mode?

File: examples/darwin-framework-tool/templates/commands.zapt [hunk 5]
```diff
@@ -222,7 +222,7 @@ public:
         [cluster subscribe{{>attribute}}WithMinInterval:[NSNumber numberWithUnsignedInt:mMinInterval] 
                                             maxInterval:[NSNumber numberWithUnsignedInt:mMaxInterval] 
                                             params:params
-                                subscriptionEstablished: nullptr reportHandler:^({{asObjectiveCClass type parent.name}} * _Nullable value, NSError * _Nullable error) {
+                                subscriptionEstablished:^(){mSubscriptionEstablished=YES;} reportHandler:^({{asObjectiveCClass type parent.name}} * _Nullable value, NSError * _Nullable error) {
```
Comment (bzbarsky-apple) [inline examples/darwin-framework-tool/templates/commands.zapt:225]:
```suggestion
                           subscriptionEstablished:^(){ mSubscriptionEstablished=YES; }
                                     reportHandler:^({{asObjectiveCClass type parent.name}} * _Nullable value, NSError * _Nullable error) {
```

**Review by bzbarsky-apple — APPROVED** [review]

File: examples/darwin-framework-tool/commands/common/CHIPCommandBridge.mm (hunk 3 above)
Comment (krypton36) [reply examples/darwin-framework-tool/commands/common/CHIPCommandBridge.mm:115]: Ok. Let me try it out with only controller startup and shutdown.

File: examples/darwin-framework-tool/commands/common/CHIPCommandBridge.mm (hunk 4 above)
Comment (krypton36) [reply examples/darwin-framework-tool/commands/common/CHIPCommandBridge.mm:114]: Can you please rephrase? I having trouble understanding this question.

File: examples/darwin-framework-tool/commands/clusters/ReportCommandBridge.h (hunk 1 above)
Comment (krypton36) [reply examples/darwin-framework-tool/commands/clusters/ReportCommandBridge.h]: Updated

File: examples/darwin-framework-tool/commands/common/CHIPCommandBridge.mm (hunk 3 above)
Comment (krypton36) [reply examples/darwin-framework-tool/commands/common/CHIPCommandBridge.mm]: Updated.

File: examples/darwin-framework-tool/templates/commands.zapt (hunk 5 above)
Comment (krypton36) [reply examples/darwin-framework-tool/templates/commands.zapt]: Updated.

File: examples/darwin-framework-tool/commands/common/CHIPCommandBridge.mm (hunk 4 above)
Comment (bzbarsky-apple) [reply examples/darwin-framework-tool/commands/common/CHIPCommandBridge.mm:114]: What calls ShutdownCommissioner() in interactive mode?  Does the "quit()" command handling do that?

Comment (krypton36) [reply examples/darwin-framework-tool/commands/common/CHIPCommandBridge.mm:114]: Yes, the "quit()" command is the only thing at the moment that calls `ShutdownCommissioner`. We may need to add signal handlers to catch and shutdown the commissioners. Should I file a followup?

Comment (bzbarsky-apple) [reply examples/darwin-framework-tool/commands/common/CHIPCommandBridge.mm:114]: No, if quit() does it we're good here.

**Review by yunhanw-google — APPROVED** [review]

**Pull request merged**